pkg/hubble/relay/pool/manager.go

```diff
@@ -0,0 +1,361 @@
+			switch cn.GetType() {
+			case peerpb.ChangeNotificationType_PEER_ADDED:
+				m.add(p)
+			case peerpb.ChangeNotificationType_PEER_DELETED:
+				m.remove(p)
+			case peerpb.ChangeNotificationType_PEER_UPDATED:
+				m.update(p)
+			}
+		}
+	}
+}
+
+func (m *Manager) manageConnections() {
+	for {
+		select {
+		case <-m.stop:
+			return
+		case name := <-m.offline:
+			m.mu.Lock()
+			p := m.peers[name]
+			m.mu.Unlock()
+			// a connection request has been made, make sure to attempt a connection
+			m.connect(p, true)
+		case <-time.After(m.opts.ConnCheckInterval):
+			var retry []*peer
+			m.mu.Lock()
+			now := time.Now()
+			for _, p := range m.peers {
+				p.mu.Lock()
+				if p.conn != nil {
+					switch p.conn.GetState() {
+					case connectivity.Connecting, connectivity.Idle, connectivity.Ready, connectivity.Shutdown:
+						p.mu.Unlock()
+						continue
+					}
+				}
+				if p.nextConnAttempt.IsZero() || p.nextConnAttempt.Before(now) {
+					retry = append(retry, p)
+				}
+				p.mu.Unlock()
+			}
+			m.mu.Unlock()
+			for _, p := range retry {
+				m.connect(p, false)
+			}
+		}
+	}
+}
+
+// Stop implements PeerManager.Stop.
+func (m *Manager) Stop() {
+	close(m.stop)
+}
+
+// List implements PeerManager.List.
+func (m *Manager) List() []Peer {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if len(m.peers) == 0 {
+		return nil
+	}
+	peers := make([]Peer, 0, len(m.peers))
+	for _, v := range m.peers {
+		// note: there shouldn't be null entries in the map
+		peers = append(peers, Peer{
+			Peer: hubblePeer.Peer{
+				Name:    v.Name,
+				Address: v.Address,
+			},
+			Conn: v.conn,
+		})
+	}
+	return peers
+}
+
+// ReportOffline implements PeerManager.ReportOffline.
+func (m *Manager) ReportOffline(name string) {
+	go func() {
+		m.mu.Lock()
+		p, ok := m.peers[name]
+		m.mu.Unlock()
+		if !ok {
+			return
+		}
+		p.mu.Lock()
+		if p.conn != nil {
+			switch p.conn.GetState() {
+			case connectivity.Connecting, connectivity.Idle, connectivity.Ready:
+				// it looks like it's actually online or being brought online
+				p.mu.Unlock()
+				return
+			}
+		}
+		p.mu.Unlock()
+		select {
+		case <-m.stop:
+		case m.offline <- name:
+		}
+	}()
+}
+
+func (m *Manager) add(hp *hubblePeer.Peer) {
+	if hp == nil {
+		return
+	}
+	p := &peer{Peer: *hp}
+	m.mu.Lock()
+	m.peers[p.Name] = p
+	m.mu.Unlock()
+	select {
+	case <-m.stop:
+	case m.offline <- p.Name:
+	}
+}
+
+func (m *Manager) remove(hp *hubblePeer.Peer) {
+	if hp == nil {
+		return
+	}
+	m.mu.Lock()
+	if p, ok := m.peers[hp.Name]; ok {
+		m.disconnect(p)
+		delete(m.peers, hp.Name)
+	}
+	m.mu.Unlock()
+}
+
+func (m *Manager) update(hp *hubblePeer.Peer) {
+	if hp == nil {
+		return
+	}
+	p := &peer{Peer: *hp}
+	m.mu.Lock()
+	if old, ok := m.peers[p.Name]; ok {
+		m.disconnect(old)
+	}
+	m.peers[p.Name] = p
+	m.mu.Unlock()
+	select {
+	case <-m.stop:
+	case m.offline <- p.Name:
+	}
+}
+
+func (m *Manager) connect(p *peer, ignoreBackoff bool) {
+	go func() {
```

Comment:
The creation of the go routine should be done by the caller to be more explicit that the `connect` is going to be done async.

Reply:
Sure, I'll change this.